pkg/http/proxy.go

```diff
@@ -31,15 +32,38 @@ func NewHeaderPruningReverseProxy(targetHost string, headersToRemove []string) *
 			req.URL.Scheme = "http"
 			req.URL.Host = targetHost
 
-			// Copied from httputil.NewSingleHostReverseProxy.
-			if _, ok := req.Header[network.UserAgentKey]; !ok {
-				// explicitly disable User-Agent so it's not set to default value
-				req.Header.Set(network.UserAgentKey, "")
-			}
+			handleHeaders(req, headersToRemove)
+		},
+	}
+}
 
-			for _, h := range headersToRemove {
-				req.Header.Del(h)
-			}
+// NewPassthroughLbHeaderPruningReverseProxy returns a httputil.ReverseProxy that proxies
+// requests to the given targetHost after removing the headersToRemove.
+// It replaces the Host header with the given spoofedHost and sets the Knative
+// passthrough-lb header to allow it to address pods in specific mesh cases.
+func NewPassthroughLbHeaderPruningReverseProxy(targetHost, spoofedHost string, headersToRemove []string) *httputil.ReverseProxy {
```

Comment:
come back java, everything is forgiven ;)

Reply:
more seriously, Im wondering if this explosion of adjectives (and the test duplication) implies we should refactor something here. One option would be e.g.

```
NewProxy(directors ...func(req *http.Request))
```

.. and then using it would be `NewProxy(PruneActivatorHeaders, AddPassthrough)`

(just one option, obviously) .

Reply:
ack, I thought about it but was lazied out. Will check again :joy:

Reply:
So one wrinkle would be that we're paying with an allocation, since the `AddPassthrough` thing would have to return a function that captures the host of each request. Meh. Shall we come up with a better name instead? :joy:

Reply:
Yeah too long of a name :|

Need `Factory` for more java feel.

Reply:
I can go with director and directorOpts

Reply:
ugh.. but isnt that already the case since we have to return a closure to capture targetHost here already?

Reply:
Right, yeah we already pay for one closure. The options would make it 2.

Reply:
Maybe we rename the default one to be `NewReverseProxy` and the new one `NewPassthroughLbReverseProxy`?

Reply:
or a NewReverseProxy with an optional `passthroughHost` param? Im ok with NewReverseProxy and NewPassthroughLbReverseProxy too, but it might be nice to avoid an if/else construction when we end up using this anyway

Reply:
The point is that each option needs to return a closure of themselves. The `PruneActivatorHeaders` thingy can be cached, but `Passthrough` needs to capture the hostOverride of each request. Can theoretically be cached across requests but... totally overkill.

Reply:
(I still think its the same, either we pass NewProxy(Prune, Passthrough(targetHost, passthroughHost)), or NewProxy((Prune, Direct(targetHost)), either way its 1 alloc, but tbh just an extra param on NewProxy or a rename is probably fine anyway)